Docs/SPE 260: update and reorganize telemetry metric references (#21965)
hashicorp/vault: website/content/docs/internals/telemetry/enable-telemetry.mdx

## Step 1: Choose an aggregation agent

@include 'telemetry/supported-aggregation-agents.mdx'

## Step 2: Enable at least one audit device

To include audit-related metrics, you must enable auditing on at least one device
with the `vault audit enable` command. For example, to enable auditing for the
`file` device and save the logs to `/var/log/vault_audit.log`:

```shell-session
$ vault audit enable file file_path=/var/log/vault_audit.log
```

By default, Enterprise installations replicate audit devices to the secondary
performance nodes in a cluster. To limit performance replication for an audit
device, use the `local` flag to mark the device as local to the current node:

```shell-session
$ vault audit enable file -local file_path=/var/log/vault_audit.log
```

## Step 3: Configure telemetry collection

To configure telemetry collection, update the telemetry stanza in your Vault
configuration with your collection preferences and aggregation agent details.

For example, the following `telemetry` stanza configures Vault with the standard
telemetry defaults and connects it to a Statsite agent running on the default
port within a company intranet at `mycompany.statsite`:

> 

Reviewer: Why do we give as our config example a stanza containing the standard defaults?  The defaults aren't necessarily "good" settings, often they're the way they are simply because we didn't want to break backwards compatibility.  Moreover, since they're the defaults, no one would ever use these settings, they'd simply omit the defaults to have a minimal config stanza, right?

Anyway, I'm not exactly objecting, just wondering why we're doing this instead of providing an example or two of typical realistic configs (maybe we do that elsewhere, like in tutorials?)

Author: It's helpful to have all the common, available parameters in an example block for a setup guide so folks know what a full telemetry stanza might look like. We do that for 2 main reasons: (1) it encourages readers to consider how they may want to customize the values right away instead of hoping they take the time to read the configuration document, and (2) it provides a clear template for them to copy/paste an change (which includes deleting things they want to leave as the default).

The stanza uses the defaults because I assumed they were reasonable values, if there are *more* reasonable values I'm happy to use those instead as a best practice.

Reviewer: > The stanza uses the defaults because I assumed they were reasonable values, if there are more reasonable values I'm happy to use those instead as a best practice.

Eh, they're not *unreasonable* for statsd.

```hcl
telemetry {
  usage_gauge_period = "10m"
  maximum_gauge_cardinality = 500
  disable_hostname = false
  enable_hostname_label = false
  lease_metrics_epsilon = "1h"
  num_lease_metrics_buckets = 168
  add_lease_metrics_namespace_labels = false
  filter_default = true
  
  statsite_address = "mycompany.statsite:8125"
}
```

<Tip heading="Use a prefix filter to reduce the volume of metrics you receive">

  Many metrics solutions charge by the metric. You can use the `prefix_filter`
  parameter to include and exclude specific values based on metric name to avoid
  paying for irrelevant information.
  
  For example, to limit your telemetry to the core token metrics plus the number
  of leases set to expire:

  ```hcl
  telemetry {

>     prefix_filter = ["+vault.token", "-vault.expire", "+vault.expire.num_leases"]

Reviewer: I haven't used this feature myself, but my read of it is that your description of what would happen is assuming filter_default=false, which is not the default.

Author: D'oh! Yeah, I forgot that part. I added the filter_default parameter to the tip to clarify

---
layout: docs
page_title: Enable Vault telemetry
description: >-
  Step-by-step guide to enabling telemetry gathering with Vault
---

# Enable Vault telemetry gathering

Collect telemetry data from your Vault installation.

## Before you start

- **You must have Vault 1.14 or later installed and running**.
- **You must have access to your [Vault configuration](/vault/docs/configuration) file**.

## Step 1: Choose an aggregation agent

@include 'telemetry/supported-aggregation-agents.mdx'

## Step 2: Enable at least one audit device

> To include audit-related metrics, you must enable auditing on at least device

Reviewer: ```suggestion
To include audit-related metrics, you must enable auditing on at least one device
```

with the `vault audit enable` command. For example, to enable auditing for the
`file` device and save the logs to `/var/log/vault_audit.log`:

```shell-session
$ vault audit enable file file_path=/var/log/vault_audit.log
```

> By default, Vault replicates audit device configuration to all nodes within a

Reviewer: ```suggestion
Vault replicates audit device configuration to all nodes within a
```

I don't think `local` changes how auditing works within a single cluster.

Author: One of the existing docs made a point of including the `-local` flag to avoid replication and implied that it applies to replication within a cluster and DR-related replication. Is it strictly DR replication that it affects?

Reviewer: My understanding is:
* within a cluster, both local and non-local audit devices are used by all nodes in the cluster when servicing requests
* across DR clusters, all local and non-local audit devices are replicated to the DR secondary
* across PR clusters, only non-local audit devices are replicated to the PR secondary (which can also have its own independent local audit devices)

Author: I tweaked the language to focus on why you might want to use the local flag. Let me know if it makes more sense.

cluster. Enterprise installations also replicate audit device configuration to

> the primary node and all secondaries in a cluster. To disable replication, use

Reviewer 2: ```suggestion
the primary node and all secondary nodes in a cluster. To disable replication, use
```

the `local` parameter:

```shell-session
$ vault audit enable file -local file_path=/var/log/vault_audit.log
```

## Step 3: Configure telemetry collection

To configure telemetry collection, update the telemetry stanza in your Vault
configuration with your collection preferences and aggregation agent details.

> For example, the following `telemetry` stanza configures Vault with the standard

Reviewer: The prefix_filter isn't "standard telemetry defaults".

Reviewer: That said, I wonder if it's worth mentioning in its own section in this page.  Many metrics solutions charge by the metric, so pointing out that it's possible to reduce the volume of metrics emitted by vault might be appreciated.

Author: Added a tip for filtering. Let me know if I'm misrepresenting anything

telemetry defaults and connects it to a Statsite agent running on the default
port within a company intranet at `mycompany.statsite`:

```json
telemetry {

>   usage_gauge_period = "10m",

Reviewer: These commas don't belong in HCL config I think, this looks like a mix of JSON and HCL formatting.

Author: The syntax should be fixed.

  maximum_gauge_cardinality = 500,
  disable_hostname = false,
  enable_hostname_label = false,
  lease_metrics_epsilon = "1h",
  num_lease_metrics_buckets = 168,
  add_lease_metrics_namespace_labels = false,
  filter_default = true,
  prefix_filter = ["+vault.token", "-vault.expire", "+vault.expire.num_leases"],
  statsite_address = "mycompany.statsite:8125"
}
```

## Step 4: Choose a reporting solution

You need to save or forward your telemetry data to a separate storage solution

> for reporting and analysis. Which solution you need depends on the feature set

Reviewer: ```suggestion
for reporting and analysis and possibly alerting. Which solution you need depends on the feature set
```

provided by your aggregation agent and the protocol support of your reporting
platform.

Popular reporting solutions compatible with Vault:

- [Grafana](https://grafana.com/grafana)
- [Graphite](https://www.hostedgraphite.com)
- [InfluxData: Telegraf](https://www.influxdata.com/time-series-platform/telegraf)
- [InfluxData: InfluxDB](https://www.influxdata.com/products/influxdb-overview)
- [InfluxData: Chronograf](https://www.influxdata.com/time-series-platform/telegraf)
- [InfluxData: Kapacitor](https://www.influxdata.com/time-series-platform/kapacitor)
- [Splunk](https://www.splunk.com)

## Next steps

- Review the full list of available
  [telemetry parameters](/vault/docs/configuration/telemetry#telemetry-parameters).

> - Review [Monitor telemetry and audit device log data](/vault/tutorials/monitoring/monitor-telemetry-audit-splunk)

Reviewer 2: ```suggestion
- Review the [Monitor telemetry and audit device log data](/vault/tutorials/monitoring/monitor-telemetry-audit-splunk)
```
